client/src/components/relatedItems/SingleCard.jsx

```diff
@@ -1,7 +1,127 @@
+  }
+  componentDidMount() {
+   this.fetchCategoryAndName();
+   this.fetchDiscountPriceAndImage();
+   this.fetchRateAndChangeToUse();
+
+  }
+
+  fetchCategoryAndName() {
+    axios({
+      method: 'get',
+      url: `/products/${this.props.id}`
+    })
+    .then((response) => {
+      this.setState({
+        category: response.data.category,
+        name: response.data.name,
+        'default_price': response.data['default_price']
+      })
+      //console.log(this.state);
+    })
+
+  }
+  fetchDiscountPriceAndImage() {
+    axios({
+      method: 'get',
+      url: `/products/${this.props.id}/styles`
+    })
+    .then((response) => {
+      var results = response.data.results;
+      //find default param is true which is the required img and price
+      var getDefault = (arr) => {
+        for(var i = 0; i < arr.length; i ++) {
+          if (arr[i]['default?'] === true) {
+            return {
+              'discount_price':results[i]['sale_price'],
+              img:results[i]['photos'][0]['url']
+            };
+          }
+        }
+      };
+      var discountAndImgObj = getDefault(results);
+      this.setState(discountAndImgObj);
+      //console.log(this.state)
+    })
+  }
+  fetchRateAndChangeToUse() {
+    axios({
+      url: `/reviews/meta?product_id=${this.props.id}`,
+      method: 'get'
+    })
+    .then((response) => {
+      var rates = response.data.ratings;
+      //console.log(rates)
+      var getAvgRateAndChangeTOOneQuarter = (rates) => {
+        var totalRates = 0;
+        var totalCounts = 0;
+        var avgRealRate = 0;
+        for(var key in rates) {
+          totalRates += key * rates[key];
+          totalCounts += rates[key] * 1;
+        }
+        //console.log(totalRates, totalCounts)
+        if (totalCounts === 0) {
+          avgRealRate = 0;
+        } else {
+          avgRealRate = totalRates / totalCounts;
+        }
+        var simpleAvgRate = 0;
+        var intPart = Math.floor(avgRealRate);
+        var decPart = avgRealRate - intPart;
+        if (decPart === 0) {
+          simpleAvgRate = intPart;
+        } else if (decPart > 0 && decPart <= 0.25) {
+          simpleAvgRate = intPart + 0.25;
+        } else if (decPart > 0.25 && decPart <= 0.5) {
+          simpleAvgRate = intPart + 0.5;
+        } else if (decPart > 0.5 && decPart <= 0.75) {
+          simpleAvgRate = intPart + 0.75;
+        } else {
+          simpleAvgRate = intPart + 1;
+        }
+        return {rate: simpleAvgRate};
+      }
+      var rateObj = getAvgRateAndChangeTOOneQuarter(rates);
+      this.setState(rateObj);
+      console.log(this.state)
+
+    })
+  }
+  render() {
+    return (
+      <div className = "singleCard">
+        <h2>singleCard</h2>
+        <p>{this.state.category}</p>
+        <p>{this.state.name}</p>
+        <p>{this.state.default_price}</p>
+        <p>{this.state.discount_price}</p>
+        <p>{this.state.img}</p>
+        <p>{this.state.rate}</p>
+      </div>
+    )
+  }
```

Comment:
You are fetching relatedItems data already in RelatedItems component, so I don't think you need to also fetch the data on each individual card. Can you instead pass all the information is as a prop instead of holding state separately here?

Reply:
I found there are quite a lot of data I need to render one card. For example, in RelatedItems component, I use /related url to only fetch the related products id. In single card component, I have to use this id to fetch price, discount price, default image and rates, they are all from different API. I'm still thinking is there a way to only fetch once and get all data. 
However, this comment is really helpful for me to think where to fetch data and how to pass though to my child component.